plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java

```diff
@@ -527,32 +529,43 @@ private Optional<String> getQualifiedParentName(MongoColumnHandle columnHandle)
      * Creates a set of sufficient columns for the input projected columns. For example,
      * if input {@param columns} include columns "a.b" and "a.b.c", then they will be projected from a single column "a.b".
      */
-    public static List<MongoColumnHandle> getSufficientColumns(List<MongoColumnHandle> columnHandles)
+    public static List<MongoColumnHandle> projectSufficientColumns(List<MongoColumnHandle> columnHandles)
     {
         List<MongoColumnHandle> sufficientColumns = new ArrayList<>();
         for (MongoColumnHandle column : columnHandles) {
-            String qualifiedName = column.getQualifiedName();
-            List<MongoColumnHandle> columnsToRemove = new ArrayList<>();
-            boolean isAddable = true;
-
-            for (MongoColumnHandle existingColumn : sufficientColumns) {
-                String name = existingColumn.getQualifiedName();
-                if (name.startsWith(qualifiedName + ".")) {
-                    columnsToRemove.add(existingColumn);
-                }
-                else if (qualifiedName.startsWith(name + ".")) {
-                    isAddable = false;
+            boolean addColumn = true;
+
+            for (int i = sufficientColumns.size() - 1; i >= 0; i--) {
+                MongoColumnHandle existingColumn = sufficientColumns.get(i);
+                if (isParentColumn(existingColumn, column)) {
+                    // If the parent of the visiting column already exists, mark it to indicate no addition is needed
+                    addColumn = false;
                     break;
                 }
+                else if (isParentColumn(column, existingColumn)) {
+                    // If the visiting column is the parent of any existing column, mark it for removal
+                    sufficientColumns.remove(i);
+                }
             }
-            sufficientColumns.removeAll(columnsToRemove);
-            if (isAddable) {
+            if (addColumn) {
                 sufficientColumns.add(column);
             }
         }
         return sufficientColumns;
     }
 
+    private static boolean isParentColumn(MongoColumnHandle parent, MongoColumnHandle child)
+    {
+        List<String> parentColumnDereferenceNames = parent.getDereferenceNames();
+        List<String> childColumnDereferenceNames = child.getDereferenceNames();
+
+        if (parent.getBaseName().equals(child.getBaseName())
+                && parentColumnDereferenceNames.size() < childColumnDereferenceNames.size()) {
+            return childColumnDereferenceNames.subList(0, parentColumnDereferenceNames.size()).containsAll(parentColumnDereferenceNames);
```

Comment:
`containsAll` doesn't ensure the order of data - like `a.b.c` and `b.a.c` would still be same right ? Applying `equals` would help us here - Can we also add test coverage for these kind of row type.